api-authentication/server/server.ts

```diff
@@ -0,0 +1,197 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+import 'dotenv/config';
+import pg from 'pg';
+import argon2 from 'argon2';
+import express from 'express';
+import jwt from 'jsonwebtoken';
+import { ClientError, errorMiddleware } from './lib/index.js';
+
+type User = {
+  userId: number;
+  username: string;
+  hashedPassword: string;
+};
+type Entry = {
+  entryId: number;
+  title: string;
+  notes: string;
+  photoUrl: string;
+};
+type Auth = {
+  username: string;
+  password: string;
+};
+
+const db = new pg.Pool({
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false,
+  },
+});
+
+const hashKey = process.env.TOKEN_SECRET;
+if (!hashKey) throw new Error('TOKEN_SECRET not found in .env');
+
+const app = express();
+app.use(express.json());
+
+app.post('/api/auth/sign-up', async (req, res, next) => {
+  try {
+    const { username, password } = req.body as Partial<Auth>;
+    if (!username || !password) {
+      throw new ClientError(400, 'username and password are required fields');
+    }
+    const hashedPassword = await argon2.hash(password);
+    const sql = `
+      insert into "users" ("username", "hashedPassword")
+      values ($1, $2)
+      returning *
+    `;
+    const params = [username, hashedPassword];
+    const result = await db.query<User>(sql, params);
+    const [user] = result.rows;
+    res.status(201).json(user);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.post('/api/auth/sign-in', async (req, res, next) => {
+  try {
+    const { username, password } = req.body as Partial<Auth>;
+    if (!username || !password) {
+      throw new ClientError(401, 'invalid login');
+    }
+    const sql = `
+      select *
+        from "users"
+        where "username" = $1;
+    `;
+    const params = [username];
+    const result = await db.query<User>(sql, params);
+    const user = result.rows[0];
+    if (!user) throw new ClientError(401, "'invalid login' error.");
+    const isMatching = await argon2.verify(user.hashedPassword, password);
+    if (!isMatching) throw new ClientError(401, "'invalid login' error.");
+    const secret = process.env.TOKEN_SECRET;
+    if (!secret) throw new Error('TOKEN_SECRET not found');
```

Comment:
This is good, but it turns out that the constant hashKey provided above already does this.